File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count
  (:require [clojure.string :as str]))

(defn count
  [nucleotide strand]
  (let [counts (frequencies strand)]
    (if (str/includes? "ACGT" (str nucleotide))
      (get counts nucleotide 0)
      (throw Throwable))))

(defn nucleotide-counts
  [strand]
  {\A (count \A strand) \T (count \T strand) \C (count \C strand) \G (count \G strand)})

File: ocaml/nucleotide-count/nucleotide_count.ml
open Core_kernel

let count strand nucleotide =
  String.count strand ~f:((=) nucleotide)

let nucleotide_counts strand =
  ['A'; 'C'; 'G'; 'T']
  |> List.filter_map ~f:(fun c -> match count strand c with
      | 0 -> None
      | n -> Some (c, n))
  |> Char.Map.of_alist_exn
